### ISSUE TITLE
Add auto-cleaning checkpoints and resuming from last.

### ISSUE THREAD

**Issue by andrewargatkiny** [andrewargatkiny/dense-attention#4]
### Checkpoints cleaning

Currently, a checkpoint is saved every `--ckpt_to_save` epochs, starting from epoch 1. During long training runs, many checkpoints get accumulated, taking much of disk space and necessitating manual clean-ups.

A better way of managing checkpoints is required. I suggest to implement the following principles:

- No checkpoints produced *before* the current launch ever get programmatically deleted. I.e. if the directory, where new checkpoints are being saved, already contains some from previous run(s), they should not be considered for managing.
- During a training run, last `n` checkpoints should be present on a running basis. E.g. if `n=3`, and the script just saved a new checkpoint, then the 4th least recent checkpoint should be deleted. `n` should be set to 3 by default.
- There should be an option to whitelist a checkpoint that gets saved every `k` epochs, so it doesn't get deleted. E.g. when `k=50`, checkpoints for epochs 1, 51, 101, 151... are not deleted regardless of current epoch number. `n` should be set to 50 by default.
- There should be an option to set an explicit whitelist with epoch numbers that should be preserved regardless of current epoch (e.g `[100, 180, 200]`).
- It would also be nice to have an option to "grant immunity" to a checkpoint that is currently best by an eval metric or any member of a set of eval metrics. Such checkpoint(s) should not be deleted until a new, better checkpoint gets produced. However, this functionality is not urgent and can be implemented later if it poses a challenge to embed into the current logic.

### Resuming from last

Currently, when one needs to resume training from a checkpoint (e.g during GLUE fine-tuning), it's necessary to launch the script with explicit provision of a checkpoint's epoch number, like:

```
configs/glue/ds_train_cola.sh --resume 3 bert_large_128_rope_loc32_swiglu_sst2
```
where `3` is an epoch number. If a user wants to resume training from the last available checkpoint, it would be more convenient to launch the script without mentioning the epoch number, e.g.:

```
configs/glue/ds_train_cola.sh --resume bert_large_128_rope_loc32_swiglu_sst2
```

Please implement this feature, for a start, in this launch script `configs/bert_relpe/ds_train_dense_attn_bert_seq128_bf16.sh`

---

Before embarking on implementation of these features, please feel free to discuss your plan and ideas in this thread.

**Comment by glaucomaa** [comment]
Plan suggestion:

Introduce a '`manage_checkpoints`' helper that is invoked right after every new `save_checkpoint`
It ignores all checkpoints that existed before the current run, keeps only the last N newly-created checkpoints, and skips any whose epoch matches the “every K epochs” rule

Add three flags (e.g. `--keep_last_ckpts`, `--protect_every_k`, `--protect_epochs`) plus a small shell-script wrapper that, when `--resume` is given without a number, finds the most recent checkpoint tag and forwards it to `deepspeed.py`

**Comment by andrewargatkiny** [comment]
Sounds good. I suggest naming the new CLI arguments `--keep_last_ckpts`, `keep_ckpt_every`, `keep_ckpt_nums` for alignment with existing naming conventions.

Please keep the new shell code in the same script for now instead of creating a wrapper. I suggest to use `--resume last` or `resume 0` or `resume -1` if it's too complicated to handle the case when there are no explicit checkpoint number after 'resume'.

Actually, I think `--resume last` is better than `--resume` as it's explicit and self-explaining.
